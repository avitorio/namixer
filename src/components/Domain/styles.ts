import styled, { css } from 'styled-components'

import { SearchResults } from 'templates/Home'
export type DomainProps = Pick<SearchResults, 'status'>

const domainModifiers = {
  available: () => css`
    opacity: 1;
  `,
  searching: () => css`
    opacity: 1;
  `,
  taken: () => css`
    background-color: #ffffff38;
    strong {
      color: #fff;
      opacity: 0.5;
    }
  `
}

export const DomainRow = styled.li<DomainProps>`
  ${({ status }) => css`
    height: 68px;
    left: 242px;
    top: 574px;
    color: #000;
    background-color: #ffffff;
    border-radius: 8px;
    max-width: 96rem;
    width: 100%;
    display: flex;
    align-items: center;
    justify-content: space-between;
    padding: 24px;
    ${!!status && domainModifiers[status]};
  `}
`

export const Domain = styled.div<DomainProps>`
  ${({ theme, status }) => css`
    strong {
      margin-right: 8px;
    }
    span {
      display: inline-flex;
      font-weight: 600;
      height: ${theme.spacings.xsmall};
      font-size: ${theme.font.sizes.xsmall};
      align-items: center;
      color: rgb(250, 250, 250);
      padding: 0rem 0.6rem;
      border-radius: 1.2rem;
      background: ${status === 'available' ? '#3cd3c1' : theme.colors.mainBg};
    }

    button {
      background: ${status === 'available'
        ? '#3cd3c1'
        : theme.colors.neutral.neutral400};
    }
  `}
`
export const Loading = styled.div`
  display: inline-block;
  width: 50px;
  height: 50px;

  &:after {
    content: ' ';
    display: flex;
    width: 34px;
    height: 34px;
    margin: 8px;
    border-radius: 50%;
    border: 3px solid #3cd3c1;
    border-color: #3cd3c1 transparent #3cd3c1 transparent;
    animation: lds-dual-ring 1.2s linear infinite;
  }
  @keyframes lds-dual-ring {
    0% {
      transform: rotate(0deg);
    }
    100% {
      transform: rotate(360deg);
    }
  }
`
