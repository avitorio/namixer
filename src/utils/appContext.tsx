import {
  Dispatch,
  SetStateAction,
  useState,
  useContext,
  createContext,
  ReactNode
} from 'react'

export interface IProviderProps {
  children?: ReactNode
}

type AppContextState = { domain: string }

type AppContextValue = {
  state: AppContextState
  setState: Dispatch<SetStateAction<AppContextState>>
}

export const AppContext = createContext<AppContextValue | undefined>(undefined)

export const AppProvider = (props: IProviderProps) => {
  const [state, setState] = useState({ domain: 'YourDomain.com' })

  return (
    <AppContext.Provider value={{ state, setState }}>
      {props.children}
    </AppContext.Provider>
  )
}

export function useAppContext() {
  const ctxValue = useContext(AppContext)
  if (ctxValue === undefined)
    throw new Error(
      'Expected an AppProvider somewhere in the react tree to set context value'
    )
  return ctxValue // now has type AppContextValue
  // or even provide domain methods for better encapsulation
}
