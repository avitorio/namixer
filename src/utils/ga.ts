// https://developers.google.com/analytics/devguides/collection/gtagjs/pages
export const pageview = (url: URL): void => {
  window.gtag('config', `${process.env.NEXT_PUBLIC_GA_TRACKING}`, {
    page_path: url
  })
}

type EventParams = { [key: string]: string | number }

type GTagEvent = {
  event: string
  params: EventParams
}

// https://developers.google.com/analytics/devguides/collection/gtagjs/events
export const event = ({ event, params }: GTagEvent): void => {
  window.gtag('event', event, {
    ...params
  })
}
