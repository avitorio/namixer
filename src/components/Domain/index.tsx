import Router from 'next/router'
import Button from 'components/Button'
import MediaMatch from 'components/MediaMatch'
import { SearchResults } from 'templates/Home'
import { useAppContext } from 'utils/appContext'
import * as S from './styles'
import { event } from 'utils/ga'

export type DomainProps = SearchResults & {
  setOpenAlert?: (open: boolean) => void
}

const Domain = ({ domain, status, setOpenAlert }: DomainProps) => {
  const appContext = useAppContext()

  const handleRegisterDomain = () => {
    appContext.setState({ domain })

    event({
      event: 'domain_register',
      params: {
        domain
      }
    })

    Router.push('/dont-you-forget-about-me')
  }

  return (
    <S.DomainRow status={status}>
      <S.Domain status={status}>
        <strong>{domain}</strong>

        <MediaMatch greaterThan="small">
          {status !== 'searching' && (
            <span>
              {status === 'taken' && <span>TAKEN</span>}
              {status === 'available' && <span>AVAILABLE</span>}
            </span>
          )}
        </MediaMatch>
      </S.Domain>
      <div>
        {status === 'taken' && (
          <Button
            size="small"
            outline
            onClick={() => setOpenAlert && setOpenAlert(true)}
          >
            Create Alert
          </Button>
        )}
        {status === 'searching' && (
          <span>
            <S.Loading />
          </span>
        )}
        {status === 'available' && (
          <Button
            size="small"
            href={`http://www.anrdoezrs.net/links/100555377/type/dlg/https://www.namecheap.com/domains/registration/results.aspx?domain=${domain}`}
            target="_blank"
            onClick={handleRegisterDomain}
          >
            Register it!
          </Button>
        )}
      </div>
    </S.DomainRow>
  )
}

export default Domain
