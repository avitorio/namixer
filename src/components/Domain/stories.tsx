import { Story, Meta } from '@storybook/react/types-6-0'
import Domain from '.'

export default {
  title: 'Domain',
  component: Domain
} as Meta

export const Default: Story = () => (
  <Domain
    {...{
      domain: 'namixer.com',
      status: 'available',
      hideTaken: false,
      setOpenAlert: () => true
    }}
  />
)
