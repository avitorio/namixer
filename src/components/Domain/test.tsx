import { AppProvider } from 'utils/appContext'
import { render } from 'utils/test-utils'

import Domain from '.'

describe('<Domain />', () => {
  it('should render the domain row', () => {
    const { container } = render(
      <AppProvider>
        <Domain
          {...{
            domain: 'namixer.com',
            status: 'available',
            setOpenAlert: () => true
          }}
        />
      </AppProvider>
    )

    expect(container.firstChild).toMatchSnapshot()
  })
})
