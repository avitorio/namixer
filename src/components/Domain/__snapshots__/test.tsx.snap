// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Domain /> should render the domain row 1`] = `
.c3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background: #E65A6D;
  color: #FAFAFA;
  font-family: Source Sans Pro,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-weight: 600;
  border: none;
  cursor: pointer;
  border-radius: 0.4rem;
  padding: 0.8rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
  height: 3.6rem;
  font-size: 1.4rem;
}

.c3:hover {
  color: #FAFAFA;
  background: #df2e46;
}

.c2 {
  display: none;
}

.c0 {
  height: 68px;
  left: 242px;
  top: 574px;
  color: #000;
  background-color: #ffffff;
  border-radius: 8px;
  max-width: 96rem;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 24px;
  opacity: 1;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 strong {
  margin-right: 8px;
}

.c1 span {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-weight: 600;
  font-size: 1.2rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: rgb(250,250,250);
  padding: 0.2rem 0.6rem;
  border-radius: 1.2rem;
  background: #3cd3c1;
}

.c1 button {
  background: #3cd3c1;
}

@media (min-width:450px) {
  .c2 {
    display: block;
  }
}

<li
  class="c0"
>
  <div
    class="c1"
  >
    <strong>
      namixer.com
    </strong>
    <div
      class="c2"
    >
      <span>
        <span>
          AVAILABLE
        </span>
      </span>
    </div>
  </div>
  <div>
    <button
      class="c3"
      href="http://www.anrdoezrs.net/links/100555377/type/dlg/https://www.namecheap.com/domains/registration/results.aspx?domain=namixer.com"
      target="_blank"
    >
      <span>
        Register it!
      </span>
    </button>
  </div>
</li>
`
